aesd-char-driver/main.c

```diff
@@ -289,6 +292,140 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count, loff
 
     return retval;
 }
+/*
+ * Description: Kernel lseek implementation
+ * file: File structure to seek on
+ * offset: File offset to seek to
+ * whence: Type of seek (SEEK_SET, SEEK_CUR, SEEK_END)
+ */
+loff_t aesd_llseek(struct file *file, loff_t offset, int whence)
+{
+	loff_t retval;
+	struct aesd_dev *char_dev = file->private_data;
+
+	// Lock the mutex with interruptible support
+	if (mutex_lock_interruptible(&aesd_device.lock)) {
+		return -ERESTARTSYS;
+	}
+
+	// Call the fixed_size_llseek function to perform the seek operation
+	retval = fixed_size_llseek(file, offset, whence, char_dev->buff_size);
+	// Check for errors returned by fixed_size_llseek
+	if (retval < 0) {
+		goto unlock;
+	}
+
+    goto unlock;
+
+unlock:
+	// Unlock the mutex before returning (including in case of errors)
+	mutex_unlock(&aesd_device.lock);
+	return retval;
+}
+
+/**
+ * Adjust the file offset (f_pos) parameter in @param filp based on the location specified by 
+ * @param write_cmd (referenced command to locate) and @param write_cmd_offset (the zero-referenced offset command to locate).
+ * @return 0 if successful, negative value if an error occurred:
+ * 
+ * - ERESTARTSYS if the mutex could not be obtained
+ * - EINVAL if write_cmd or write_cmd_offset was out of range
+ */
+static long aesd_adjust_file_offset(struct file *filp, unsigned int write_cmd, unsigned int write_cmd_offset)
+{
+	long retval = 0;
+	struct aesd_dev *char_dev = filp->private_data;
+	loff_t updated_fpos_offset = 0;
+
+	// Lock the mutex to ensure exclusive access to device data
+	if (mutex_lock_interruptible(&aesd_device.lock)) {
+		return -ERESTARTSYS;
+	}
+
+	// Check for valid write_cmd
+	if (write_cmd >= AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED)
```

Comment:
@Suraj-Ajjampur does this check need to be after the mutex lock?

Reply:
Since write_cmd is an input parameter and its validity isn't dependent on any shared resource (assuming AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED is a constant and not a shared resource subject to change), checking it before acquiring the mutex does not compromise thread safety.